src/xenia/kernel/xboxkrnl/xboxkrnl_hal.cc

```diff
@@ -17,16 +17,29 @@ namespace xe {
 namespace kernel {
 namespace xboxkrnl {
 
-void HalReturnToFirmware_entry(dword_t routine) {
-  // void
-  // IN FIRMWARE_REENTRY  Routine
-
-  // Routine must be 1 'HalRebootRoutine'
-  assert_true(routine == 1);
+constexpr std::array<const char*, 9> FirmwareReentryMessage = {
+    "hard poweroff",
+    "hard reset (video error)",
+    "hard reset (used for dumpwritedump/frozen processor)",
+    "hard reset",
+    "power off (hard)",
+    "power off (nice)",
+    "Shut off (Lost Settings)",
+    "Shut off (Frozen Console)",
+    "Shut off",
+};
 
+void HalReturnToFirmware_entry(dword_t routine) {
   // TODO(benvank): diediedie much more gracefully
   // Not sure how to blast back up the stack in LLVM without exceptions, though.
-  XELOGE("Game requested shutdown via HalReturnToFirmware");
+  std::string shut_off_request = "Game requested a ";
```

Comment:
You can do something fancy:

```
constexpr std::string_view halMessageFormat = "Game requested a {} via HalReturnToFirmware";
constexpr std::string_view halUnknownMessageFormat = "Reboot (Routine Code: {})";
```

and then everything can be simplified to: 

```
const std::string exitMessage = fmt::format(
      halMessageFormat,
      static_cast<size_t>(routine) < FirmwareReentryMessage.size()
          ? FirmwareReentryMessage[routine]
          : fmt::format(std::string(halUnknownMessageFormat), routine.value()));
```

Reply:
Sure, but if I can avoid manual string adding I will try to do it. I know it will be faster, but let's be real. We don't need performance here